lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,340 @@
+          post[:card][:present] = options[:present]
+          post[:card][:securityCode] = {}
+        else
+          post[:card] = {} if post[:card].nil?
+          post[:card][:token] = {}
+          post[:card][:token][:value] = credit_card
+        end
+        if options[:card].present?
+          post[:card][:expirationDate] = options[:card][:expiration_date] if options[:card][:expiration_date]
+          post[:card][:number] = options[:card][:number] if options[:card][:number].present?
+          post[:card][:entryMode] = options[:card][:entry_mode]
+          post[:card][:present] = options[:card][:present]
+          post[:card][:securityCode] = {} unless post[:card][:securityCode]
+          post[:card][:securityCode][:indicator] = options[:card][:indicator] if options[:card][:indicator]
+          post[:card][:securityCode][:value] = options[:card][:value] if options[:card][:value]
+          post[:card][:securityCode] = {}
+        end
+        if options[:security_code].present?
+          post[:card][:securityCode] = {}
+          post[:card][:securityCode][:indicator] = options[:security_code][:indicator]
+          post[:card][:securityCode][:value] = options[:security_code][:value]
+        end
+      end
+
+      def add_customer(post, options)
+        if address = options[:customer]
+          post[:customer] = {}
+          post[:customer][:addressLine1] = address[:address_line1] if address[:address_line1]
+          post[:customer][:firstName] = address[:first_name]
+          post[:customer][:lastName] = address[:last_name]
+          post[:customer][:postalCode] = address[:postal_code]
+        end
+      end
+
+      def add_purchase_card(post, options)
+        post[:transaction] = {}
+        post[:transaction][:invoice] = options[:invoice]
+        post[:transaction][:notes] = options[:notes] if options[:notes].present?
+        if options[:purchase_card].present?
+          post[:transaction][:purchaseCard] = {}
+          post[:transaction][:purchaseCard][:customerReference] = options[:purchase_card][:customer_reference]
+          post[:transaction][:purchaseCard][:destinationPostalCode] = options[:purchase_card][:destination_postal_code]
+          post[:transaction][:purchaseCard][:productDescriptors] = options[:purchase_card][:product_descriptors]
+        end
+      end
+
+      def add_override_bus_date(post, options)
+        post[:overrideBusDate] = options[:override_bus_date] if options[:override_bus_date].present?
+      end
+
+      def add_amex(post, options)
+        if options[:amex].present?
+          post[:amex] = {}
+          post[:amex][:propertyCode] = options[:amex][:property_code]
+        end
+      end
+
+      def add_auto(post, options)
+        if options[:auto].present?
+          post[:auto] = {}
+          post[:auto][:estimatedDays] = options[:auto][:estimated_days]
+        end
+      end
+
+      def add_card_on_file(post, options)
```

Comment:
This still needs to be updated to work with [our Stored Credentials framework](https://spreedly.atlassian.net/wiki/spaces/PS/pages/1358856480/Shift4+Technical+Requirements#Spreedly-Stored-Credentials-Integration).